connman/src/service.c

```diff
@@ -5018,9 +5029,28 @@ static DBusMessage *set_property(DBusConnection *conn,
 
 		g_string_free(str, TRUE);
 
-		if (gw && strlen(gw))
+		if (gw && strlen(gw)) {
+			int gw_family = connman_inet_check_ipaddress(gw);
+
 			__connman_service_nameserver_add_routes(service, gw);
 
+			/*
+			 * After the user set ns routes have been added restore
+			 * old refcounters to avoid situation where online
+			 * check might remove the routes if this change happens
+			 * during the initial resolving of the online check URL.
+			 * Set the old value only for the IP type for which the
+			 * routes were added.
+			 */
+			if (ipv4_refcount_old && gw_family == AF_INET)
+				service->nameservers_ipv4_refcount =
```

Comment:
If I understand what is going on...
- at start: cache original refcounts
- earlier: ensure that both refcounts=1 or 0 + del_routes() -> both refcounts==0
- now, right after add_routes(): either ipv4 or ipv6 refcount == 1 (and the other is zero)
- we want that the refcount that is non-zero to be restored to original value

Now, I wonder why one would use test that feels like it is based on assumptions on what ought to have happened instead of just checking what happened? I mean, could this be simply:

    if (service->nameservers_ipvN_refcount > 0)
        service->nameservers_ipvN_refcount = ipvN_refcount_old;

Reply:
This is because the type of the nameserver routes added are determined by the `gw` family, here the `gw_family` type. Well, thinking that way cleans up the code and ends up reliable result. Good suggestion, I'll change that.

Reply:
Changed with b38de1d56f3095940f4a92f0eb5709bd086609de, which is going to be squeezed into some other commit later.